runtime/j9vm/jvm.c

```diff
@@ -4542,54 +4616,64 @@ JVM_RegisterSignal(jint sigNum, void* handler)
 	memset(&oldSignalAction, 0, sizeof(struct sigaction));
 #endif /* !defined(WIN32) */
 
-	Trc_SC_RegisterSignal();
+	Trc_SC_RegisterSignal_Entry(sigNum, handler);
 
-	if (isSignalUsedByVM(sigNum)) {
-		/* Don't allow user to register a native handler since
-		 * the signal is already used by the VM.
-		 */
-	} else if (J9_ARE_NO_BITS_SET(javaVM->sigFlags, J9_SIG_XRS_SYNC)
-			&& isSignalSpecial(sigNum)) {
-		/* Don't allow user to register a native handler since
-		 * the signal is already used by the VM.
+	if (isSignalReservedByJVM(sigNum)) {
+		/* Return error if signal is reserved by the JVM. */
+		return (void *)J9_SIG_ERR;
+	} else if (J9_ARE_ANY_BITS_SET(javaVM->sigFlags, J9_SIG_XRS_ASYNC) && isShutdownSignal) {
+		/* Don't register a handler for shutdown signals if -Xrs or -Xrs:async is specified.
+		 * If -Xrs:sync is specified, then register a handler for shutdown signals.
 		 */
+		return (void *)J9_SIG_ERR;
+	} else if (isShutdownSignal && isSignalIgnoredByOS(sigNum)) {
+		/* Ignore shutdown signal if it is ignored. */
+		return (void *)J9_SIG_IGNORED;
 	} else {
-		/* Register the signal */
-#if defined(WIN32)
+		/* Register the signal. */
 		if ((void *)J9_PRE_DEFINED_HANDLER_CHECK == handler) {
-			handler = (void *)dummySignalHandler;
-		}
-		oldHandler = OMRSIG_SIGNAL(sigNum, handler);
-#else /* defined(WIN32) */
-		sigemptyset(&newSignalAction.sa_mask);
-#if !defined(J9ZTPF)
-		newSignalAction.sa_flags = SA_RESTART;
-#else /* !defined(J9ZTPF) */
-		newSignalAction.sa_flags = 0;
-#endif /* !defined(J9ZTPF) */
-		if ((void *)J9_PRE_DEFINED_HANDLER_CHECK == handler) {
-			newSignalAction.sa_handler = dummySignalHandler;
+			if (0 != javaVM->internalVMFunctions->registerPredefinedHandler(javaVM, sigNum, &oldHandler)) {
+				Trc_SC_RegisterSignal_FailedToRegisterHandler(sigNum, handler, oldHandler);
+			}
 		} else {
-			newSignalAction.sa_handler = (void (*)(int))handler;
-		}
-		OMRSIG_SIGACTION(sigNum, &newSignalAction, &oldSignalAction);
+#if defined(WIN32)
+			oldHandler = (void *)OMRSIG_SIGNAL(sigNum, handler);
+#else /* defined(WIN32) */
+			/* Don't block any signals. */
+			if (0 != sigemptyset(&newSignalAction.sa_mask)) {
+				return (void *)J9_SIG_ERR;
```

Comment:
This return is missing a TraceExit

Reply:
fixed. added an `exit` label. `oldHandler` is initialized to `J9_SIG_ERR` so we can `goto exit` and return `oldHandler`.